### PR TITLE
Add /milestones endpoint

### PR DESCRIPTION
**Pull request by bhalbright**
* Creating a /milestones endpoint which basically serves as a dashboard view for milestones, very similar to the /issues or /pulls page.

Ref issue #8232

### CODE REVIEW

**Comment by bhalbright** [comment]
This is ready for review

**Comment by lunny** [comment]
Please add a configura to enable/disable the endpoint.

**Comment by 6543** [comment]
@bhalbright can you add a screenshot how this would look like - thx

**Comment by 6543** [comment]
> Please add a configura to enable/disable the endpoint.

@lunny should we add an setting section callet `Unit` wich has all thigs in it about enable/disable UnitType's globaly ?

so setting.Unit.KanbanBoard; setting.Unit.Milestone ... can be disabled ?

**Comment by bhalbright** [comment]
> Please add a configura to enable/disable the endpoint.

@lunny I added a setting to enable/disable this feature, it is enabled by default. Hope this is what you had in mind, happy to change if not.

**Comment by bhalbright** [comment]
> @bhalbright can you add a screenshot how this would look like - thx

@6543 I posted an animated gif in the issue page: https://github.com/go-gitea/gitea/issues/8232#issuecomment-547721687. If you want to see anything else let me know, thanks!

**Comment by lunny** [comment]
@bhalbright Thanks! That's what I said. @6543 I don't mean a per-repository setting.

File: models/issue_milestone.go [hunk 1]
```diff
@@ -465,3 +493,83 @@ func DeleteMilestoneByRepoID(repoID, id int64) error {
 	}
 	return sess.Commit()
 }
+
+// CountMilestonesByRepo map from repoIDs to number of milestones matching the options`
+func CountMilestonesByRepo(repoIDs []int64, isClosed bool) (map[int64]int64, error) {
```
Comment (6543) [inline models/issue_milestone.go]: Smal nit: this func consume RepoIDs but only next function has ByRepoIDs in name ...

Comment (6543) [reply models/issue_milestone.go]: Edit: current existing func already mix this :( so wuld say optional ...

Comment (bhalbright) [reply models/issue_milestone.go]: I think that's a fair point, I changed to "CountMilestonesByRepoIDs" so the naming matches up with the next function "GetMilestonesByRepoIDs"

Comment (6543) [reply models/issue_milestone.go]: looks good, I'll try it later :)

**Review by lunny — APPROVED** [review]

**Review by lunny — APPROVED** [review]

**Comment by 6543** [comment]
:eyes:

**Comment by 6543** [comment]
can you make multible repos klikable like it is now for issue page ?
and the number of milestone should not change if you have klicked a repo

**Comment by lunny** [comment]
@6543 that could be another PR if there are many changes.

**Comment by bhalbright** [comment]
@6543 @lunny  Thanks, I didn't notice that PR, I had done most of the code before that went in. I started looking at it some tonight, it doesn't look that bad. But I only have maybe 30-45 minutes a day to work on open source so it may take me a few days before I have it ready. Thanks.

**Comment by 6543** [comment]
@bhalbright can you resolve conflict (caused by 484e3dc5a9012a50838bf8d00de3ffd136c0759b)

**Comment by 6543** [comment]
> and the number of milestone should not change if you have klicked a repo

@bhalbright the diff to add:

```diff
diff --git a/models/issue_milestone.go b/models/issue_milestone.go
index a27c797b3..c828f6e13 100644
--- a/models/issue_milestone.go
+++ b/models/issue_milestone.go
@@ -548,23 +548,18 @@ type UserMilestoneStats struct {
        OpenCount, ClosedCount int64
 }
 
-// GetUserMilestoneStats returns milestone statistic information for dashboard by given conditions.
-func GetUserMilestoneStats(userID int64, repoID int64, userRepoIDs []int64) (*UserMilestoneStats, error) {
+// GetMilestoneStats returns milestone statistic information for dashboard by given conditions.
+func GetMilestoneStats(userRepoIDs []int64) (*UserMilestoneStats, error) {
        var err error
        stats := &UserMilestoneStats{}
 
-       cond := builder.NewCond()
-       if repoID > 0 {
-               cond = cond.And(builder.Eq{"repo_id": repoID})
-       }
-
-       stats.OpenCount, err = x.Where(cond).And("is_closed = ?", false).
+       stats.OpenCount, err = x.Where("is_closed = ?", false).
                And(builder.In("repo_id", userRepoIDs)).
                Count(new(Milestone))
        if err != nil {
                return nil, err
        }
-       stats.ClosedCount, err = x.Where(cond).And("is_closed = ?", true).
+       stats.ClosedCount, err = x.Where("is_closed = ?", true).
                And(builder.In("repo_id", userRepoIDs)).
                Count(new(Milestone))
        if err != nil {
diff --git a/routers/user/home.go b/routers/user/home.go
index 978f819c0..dcf7e4dc6 100644
--- a/routers/user/home.go
+++ b/routers/user/home.go
@@ -285,9 +285,9 @@ func Milestones(ctx *context.Context) {
                }
        }
 
-       milestoneStats, err := models.GetUserMilestoneStats(ctxUser.ID, repoID, userRepoIDs)
+       milestoneStats, err := models.GetMilestoneStats(userRepoIDs)
        if err != nil {
-               ctx.ServerError("GetUserMilestoneStats", err)
+               ctx.ServerError("GetMilestoneStats", err)
                return
        }
 
diff --git a/templates/user/dashboard/milestones.tmpl b/templates/user/dashboard/milestones.tmpl
index ae99c1b98..836c12349 100644
--- a/templates/user/dashboard/milestones.tmpl
+++ b/templates/user/dashboard/milestones.tmpl
@@ -5,7 +5,7 @@
                <div class="ui stackable grid">
                        <div class="four wide column">
                                <div class="ui secondary vertical filter menu">
-                                       <a class="item" href="{{.Link}}?type=your_repositories&repo={{.RepoID}}&sort={{$.SortType}}&state={{.State}}">
+                                       <a class="item" href="{{.Link}}?type=your_repositories&sort={{$.SortType}}&state={{.State}}">
                                                {{.i18n.Tr "home.issues.in_your_repos"}}
                                                <strong class="ui right">{{.Total}}</strong>
                                        </a>

```

**Comment by 6543** [comment]
> and the number of milestone should not change if you have klicked a repo

@bhalbright the diff to add:

```diff
diff --git a/models/issue_milestone.go b/models/issue_milestone.go
index a27c797b3..5fd1e1b8c 100644
--- a/models/issue_milestone.go
+++ b/models/issue_milestone.go
@@ -543,28 +543,23 @@ func GetMilestonesByRepoIDs(repoIDs []int64, page int, isClosed bool, sortType s
        return miles, sess.Find(&miles)
 }
 
-// UserMilestoneStats represents milestone statistic information.
-type UserMilestoneStats struct {
+// MilestonesStats represents milestone statistic information.
+type MilestonesStats struct {
        OpenCount, ClosedCount int64
 }
 
-// GetUserMilestoneStats returns milestone statistic information for dashboard by given conditions.
-func GetUserMilestoneStats(userID int64, repoID int64, userRepoIDs []int64) (*UserMilestoneStats, error) {
+// GetMilestonesStats returns milestone statistic information for dashboard by given conditions.
+func GetMilestonesStats(userRepoIDs []int64) (*MilestonesStats, error) {
        var err error
-       stats := &UserMilestoneStats{}
+       stats := &MilestonesStats{}
 
-       cond := builder.NewCond()
-       if repoID > 0 {
-               cond = cond.And(builder.Eq{"repo_id": repoID})
-       }
-
-       stats.OpenCount, err = x.Where(cond).And("is_closed = ?", false).
+       stats.OpenCount, err = x.Where("is_closed = ?", false).
                And(builder.In("repo_id", userRepoIDs)).
                Count(new(Milestone))
        if err != nil {
                return nil, err
        }
-       stats.ClosedCount, err = x.Where(cond).And("is_closed = ?", true).
+       stats.ClosedCount, err = x.Where("is_closed = ?", true).
                And(builder.In("repo_id", userRepoIDs)).
                Count(new(Milestone))
        if err != nil {
diff --git a/routers/user/home.go b/routers/user/home.go
index 978f819c0..c68d8421e 100644
--- a/routers/user/home.go
+++ b/routers/user/home.go
@@ -285,9 +285,9 @@ func Milestones(ctx *context.Context) {
                }
        }
 
-       milestoneStats, err := models.GetUserMilestoneStats(ctxUser.ID, repoID, userRepoIDs)
+       milestoneStats, err := models.GetMilestonesStats(userRepoIDs)
        if err != nil {
-               ctx.ServerError("GetUserMilestoneStats", err)
+               ctx.ServerError("GetMilestoneStats", err)
                return
        }
 
diff --git a/templates/user/dashboard/milestones.tmpl b/templates/user/dashboard/milestones.tmpl
index ae99c1b98..836c12349 100644
--- a/templates/user/dashboard/milestones.tmpl
+++ b/templates/user/dashboard/milestones.tmpl
@@ -5,7 +5,7 @@
                <div class="ui stackable grid">
                        <div class="four wide column">
                                <div class="ui secondary vertical filter menu">
-                                       <a class="item" href="{{.Link}}?type=your_repositories&repo={{.RepoID}}&sort={{$.SortType}}&state={{.State}}">
+                                       <a class="item" href="{{.Link}}?type=your_repositories&sort={{$.SortType}}&state={{.State}}">
                                                {{.i18n.Tr "home.issues.in_your_repos"}}
                                                <strong class="ui right">{{.Total}}</strong>
                                        </a>
```

**Comment by 6543** [comment]
and test fix:
```diff
diff --git a/models/issue_milestone_test.go b/models/issue_milestone_test.go
index 71ff9ed5b..787b849cc 100644
--- a/models/issue_milestone_test.go
+++ b/models/issue_milestone_test.go
@@ -364,14 +364,13 @@ func TestLoadTotalTrackedTime(t *testing.T) {
        assert.Equal(t, milestone.TotalTrackedTime, int64(3662))
 }
 
-func TestGetUserMilestoneStats(t *testing.T) {
+func TestGetMilestonesStats(t *testing.T) {
        assert.NoError(t, PrepareTestDatabase())
        repo1 := AssertExistsAndLoadBean(t, &Repository{ID: 1}).(*Repository)
        repo2 := AssertExistsAndLoadBean(t, &Repository{ID: 2}).(*Repository)
-       user := AssertExistsAndLoadBean(t, &User{ID: 2}).(*User)
 
-       milestoneStats, err := GetUserMilestoneStats(user.ID, repo1.ID, []int64{repo1.ID, repo2.ID})
+       milestoneStats, err := GetMilestonesStats([]int64{repo1.ID, repo2.ID})
        assert.NoError(t, err)
-       assert.EqualValues(t, repo1.NumOpenMilestones, milestoneStats.OpenCount)
-       assert.EqualValues(t, repo1.NumClosedMilestones, milestoneStats.ClosedCount)
+       assert.EqualValues(t, repo1.NumOpenMilestones+repo2.NumOpenMilestones, milestoneStats.OpenCount)
+       assert.EqualValues(t, repo1.NumClosedMilestones+repo2.NumClosedMilestones, milestoneStats.ClosedCount)
 }

```

**Comment by 6543** [comment]
@bhalbright do you know how to **apply** this diffs? (- if you are ok with them)

**Comment by 6543** [comment]
and add integration test

```diff
diff --git a/integrations/links_test.go b/integrations/links_test.go
index fc0f16455..db86b2281 100644
--- a/integrations/links_test.go
+++ b/integrations/links_test.go
@@ -86,6 +86,12 @@ func testLinksAsUser(userName string, t *testing.T) {
                "/pulls?type=your_repositories&repos=[0]&sort=&state=closed",
                "/pulls?type=assigned&repos=[0]&sort=&state=closed",
                "/pulls?type=created_by&repos=[0]&sort=&state=closed",
+               "/milestones",
+               "/milestones?repo=0&sort=mostcomplete&state=closed",
+               "/milestones?type=your_repositories&sort=mostcomplete&state=closed",
+               "/milestones?sort=&repo=1&state=closed",
+               "/milestones?sort=&repo=1&state=open",
+               "/milestones?repo=0&sort=mostissues&state=open",
                "/notifications",
                "/repo/create",
                "/repo/migrate",
```

**Comment by 6543** [comment]
@bhalbright you can pikup all commits on https://github.com/6543-forks/gitea/tree/milestones-multiselectable

I made it multi selectable also: eb674fcaea0cfcf6a86255a15c1110a8841c70aa

File: integrations/links_test.go [hunk 2]
```diff
@@ -86,6 +86,12 @@ func testLinksAsUser(userName string, t *testing.T) {
 		"/pulls?type=your_repositories&repos=[0]&sort=&state=closed",
 		"/pulls?type=assigned&repos=[0]&sort=&state=closed",
 		"/pulls?type=created_by&repos=[0]&sort=&state=closed",
+		"/milestones",
+		"/milestones?repo=0&sort=mostcomplete&state=closed",
+		"/milestones?type=your_repositories&sort=mostcomplete&state=closed",
+		"/milestones?sort=&repo=1&state=closed",
```
Comment (6543) [inline integrations/links_test.go]:
```suggestion
		"/milestones?sort=&repos=[1]&state=closed",
```

File: integrations/links_test.go [hunk 3]
```diff
@@ -86,6 +86,12 @@ func testLinksAsUser(userName string, t *testing.T) {
 		"/pulls?type=your_repositories&repos=[0]&sort=&state=closed",
 		"/pulls?type=assigned&repos=[0]&sort=&state=closed",
 		"/pulls?type=created_by&repos=[0]&sort=&state=closed",
+		"/milestones",
+		"/milestones?repo=0&sort=mostcomplete&state=closed",
```
Comment (6543) [inline integrations/links_test.go]:
```suggestion
		"/milestones?sort=mostcomplete&state=closed",
```

File: integrations/links_test.go [hunk 4]
```diff
@@ -86,6 +86,12 @@ func testLinksAsUser(userName string, t *testing.T) {
...
+		"/milestones?type=your_repositories&sort=mostcomplete&state=closed",
+		"/milestones?sort=&repo=1&state=closed",
+		"/milestones?sort=&repo=1&state=open",
```
Comment (6543) [inline integrations/links_test.go]:
```suggestion
		"/milestones?sort=&repos=[1]&state=open",
```

File: integrations/links_test.go [hunk 5]
```diff
@@ -86,6 +86,12 @@ func testLinksAsUser(userName string, t *testing.T) {
...
+		"/milestones?repo=0&sort=mostissues&state=open",
```
Comment (6543) [inline integrations/links_test.go]:
```suggestion
		"/milestones?repos=[0]&sort=mostissues&state=open",
```

**Review by 6543 — APPROVED** [review]
only adjust integration :)

beside that OK

**Comment by 6543** [comment]
@bhalbright sory i forgod to change test after multi select ...

**Comment by bhalbright** [comment]
@6543 thanks for your help! I tested some locally and it looks good to me too. Thanks again!

**Review by zeripath — CHANGES_REQUESTED** [review]
Need to update the testing urls as per @6543

**Comment by bhalbright** [comment]
> Need to update the testing urls as per @6543

@zeripath  @6543  Done, thanks

**Review by zeripath — APPROVED** [review]

**Comment by zeripath** [comment]
make lg-tm work

**Pull request merged**